s3migrate/shared.py

```diff
@@ -0,0 +1,397 @@
+
+    Returns a hex string.
+    """
+    return hash_fileobj(fileobj, hashlib.md5()).hexdigest()
+
+
+def md5_fileobj_b64(fileobj):
+    """Efficiently calculate the MD5 sum for a file object.
+
+    Returns a Base64 encoded string.
+    """
+    return base64.b64encode(
+        hash_fileobj(fileobj, hashlib.md5()).digest()
+    ).decode("utf-8")
+
+
+def sha1_fileobj_hex(fileobj):
+    """Efficiently calculate the SHA1 sum for a file object.
+
+    Returns a hex string.
+    """
+    return hash_fileobj(fileobj, hashlib.sha1()).hexdigest()
+
+
+def sha1_fileobj_b64(fileobj):
+    """Efficiently calculate the SHA1 sum for a file object.
+
+    Returns a Base64 encoded string.
+    """
+    return base64.b64encode(
+        hash_fileobj(fileobj, hashlib.sha1()).digest()
+    ).decode("utf-8")
+
+
+class MogileFile():
+    """Represents a file stored in mogile."""
+
+    def __init__(self, fid: int, dkey: str, length: int):
+        """Initialize a MogileFile.
+
+        Arguments are as stored in the mogile `file` database table.
+        """
+        self.length = length
+        self.fid = fid
+        self.dkey = dkey
+        if dkey[36:] == ".tmp":
+            # These seem to be old junk leftover from failed ingests.
+            # Check later.
+            self.temp = True
+            self.sha1sum = None
+            self.mogile_bucket = None
+        else:
+            self.temp = False
+            self.sha1sum = dkey[0:40]
+            self.mogile_bucket = dkey[41:]
+            assert len(self.sha1sum) == 40
+            assert dkey[40] == "-"
+            assert len(self.mogile_bucket) > 1
+
+    def __eq__(self, other):
+        """Equality check."""
+        if not isinstance(other, MogileFile):
+            # Delegate comparison to the other instance's __eq__.
+            return NotImplemented
+        return(self.dkey == other.dkey and
+               self.fid == other.fid and
+               self.length == other.length)
+
+    @classmethod
+    def parse_row(cls, row: dict):
+        """Take a file row and return a MogileFile."""
+        # (fid, dmid, dkey, length, classid, devcount)
+        # Sanity check, we only use one "domain" and one class
+        assert row[1] == 1, "Bad domain"
+        assert row[4] == 0, "Bad class"
+        return cls(dkey=row[2],
+                   fid=row[0],
+                   length=row[3])
+
+    def exists_in_bucket(self, client, s3_bucket, key):
+        """Check if object with key is in the bucket.
+
+        Returns False if object is not present, otherwise the md5string.
+        """
+        obj = client.Object(s3_bucket, key)
+        try:
+            obj.load()
+            if obj.content_length != self.length:
+                return False
+        except ClientError as ex:
+            if ex.response['Error']['Code'] == "404":
+                return False
+            raise
+        return obj.e_tag.replace("\"", "")  # Why does AWS include this?
```

Comment:
how annoying. I guess that's just the way it is https://github.com/aws/aws-sdk-net/issues/815

Reply:
Interesting! Yes, it's weird.